File: backend/CampusMart/technominds/src/main/java/com/appdevg5/technominds/Review/ReviewService.java
package com.appdevg5.technominds.Review;

import com.appdevg5.technominds.Order.OrderEntity;
import com.appdevg5.technominds.Order.OrderRepository;
import com.appdevg5.technominds.Profile.ProfileRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;
import java.util.Optional;

/**
 * Service layer for managing Review-related business logic.
 */
@Service
public class ReviewService {

    private final ReviewRepository reviewRepository;
    private final ProfileRepository profileRepository;
    private final OrderRepository orderRepository;

    public ReviewService(ReviewRepository reviewRepository, ProfileRepository profileRepository, OrderRepository orderRepository) {
        this.reviewRepository = reviewRepository;
        this.profileRepository = profileRepository;
        this.orderRepository = orderRepository;
    }

    // READ
    public Optional<ReviewEntity> getReviewById(Integer id) {
        return reviewRepository.findById(id);
    }

    // Reviews received by a seller (previously called reviewee)
    public List<ReviewEntity> getReviewsBySellerId(Integer sellerId) {
        return reviewRepository.findBySeller_IdOrderByCreatedAtDesc(sellerId);
    }

    // Reviews written by a specific reviewer
    public List<ReviewEntity> getReviewsByReviewerId(Integer reviewerId) {
        return reviewRepository.findByReviewer_IdOrderByCreatedAtDesc(reviewerId);
    }

    // Reviews for a specific product (optional relation)
    public List<ReviewEntity> getReviewsByProductId(Integer productId) {
        return reviewRepository.findByProduct_IdOrderByCreatedAtDesc(productId);
    }

    // CREATE
    @Transactional
    public ReviewEntity createReview(ReviewEntity review) {
        // *** IMPORTANT BUSINESS LOGIC NOTE ***
        // 1. Verify that the Order linked (if any) is in a 'COMPLETED' state.
        // 2. Verify reviewer and seller relationship against the order.
        // 3. The DB unique constraint should prevent duplicate reviews for the same order.
        return reviewRepository.save(review);
    }

    // CREATE with validation
    @Transactional
    public ReviewEntity createReviewWithValidation(ReviewEntity review) {
        // Validation 1: If order is provided, verify it exists and is completed
        if (review.getOrder() != null) {
            Integer orderId = review.getOrder().getId();
            Optional<OrderEntity> orderOpt = orderRepository.findById(orderId);
            
            if (orderOpt.isEmpty()) {
                throw new IllegalArgumentException("Order not found with ID: " + orderId);
            }
            
            OrderEntity order = orderOpt.get();
            
            // Check if order is completed
            if (!"completed".equalsIgnoreCase(order.getStatus())) {
                throw new IllegalArgumentException("Reviews can only be submitted for completed orders. Current status: " + order.getStatus());
            }
            
            // Validation 2: Check if review already exists for this order
            List<ReviewEntity> existingReviews = reviewRepository.findByOrder_Id(orderId);
            if (!existingReviews.isEmpty()) {
                throw new IllegalArgumentException("A review already exists for this order");
            }
            
            // Validation 3: Verify reviewer is the buyer of the order
            if (!order.getBuyer().getId().equals(review.getReviewer().getId())) {
                throw new IllegalArgumentException("Only the buyer of the order can submit a review");
            }
            
            // Validation 4: Verify seller matches the order
            if (!order.getSeller().getId().equals(review.getSeller().getId())) {
                throw new IllegalArgumentException("Seller mismatch with order");
            }
        }
        
        // Save the review
        ReviewEntity savedReview = reviewRepository.save(review);
        
        // Auto-update seller's rating
        updateSellerRating(review.getSeller().getId());
        
        return savedReview;
    }

    // READ all reviews by sellerId (reviews this seller received)
    public List<ReviewEntity> getReviewsBySeller(Integer sellerId) {
        return reviewRepository.findBySeller_IdOrderByCreatedAtDesc(sellerId);
    }

    // READ all reviews written by this reviewer
    public List<ReviewEntity> getReviewsWritten(Integer reviewerId) {
        return reviewRepository.findByReviewer_IdOrderByCreatedAtDesc(reviewerId);
    }

    // UPDATE
    @Transactional
    public Optional<ReviewEntity> updateReview(Integer id, ReviewEntity reviewDetails) {
        return reviewRepository.findById(id).map(existingReview -> {
            // Only allow updating comment and rating
            existingReview.setRating(reviewDetails.getRating());
            existingReview.setComment(reviewDetails.getComment());
            return reviewRepository.save(existingReview);
        });
    }

    // DELETE
    @Transactional
    public void deleteReview(Integer id) {
        reviewRepository.deleteById(id);
    }
    
    /**
     * Get detailed reviews for a seller with reviewer information and pagination.
     * @param sellerId The seller's profile ID
     * @param page Page number (0-indexed)
     * @param size Number of reviews per page
     * @param sortBy Sort field (recent, highest, lowest)
     * @return Page of ReviewDetailDTO with reviewer details
     */
    @Transactional(readOnly = true)
    public Page<ReviewDetailDTO> getDetailedReviewsBySeller(Integer sellerId, int page, int size, String sortBy) {
        Sort sort;
        switch (sortBy) {
            case "highest":
                sort = Sort.by(Sort.Direction.DESC, "rating").and(Sort.by(Sort.Direction.DESC, "createdAt"));
                break;
            case "lowest":
                sort = Sort.by(Sort.Direction.ASC, "rating").and(Sort.by(Sort.Direction.DESC, "createdAt"));
                break;
            case "recent":
            default:
                sort = Sort.by(Sort.Direction.DESC, "createdAt");
                break;
        }
        
        Pageable pageable = PageRequest.of(page, size, sort);
        Page<ReviewEntity> reviewsPage = reviewRepository.findBySeller_Id(sellerId, pageable);
        
        return reviewsPage.map(review -> {
            // Force lazy loading of relationships
            review.getReviewer().getId();
            review.getSeller().getId();
            if (review.getProduct() != null) {
                review.getProduct().getId();
            }
            
            return new ReviewDetailDTO(
                review.getId(),
                review.getReviewer().getId(),
                review.getReviewer().getFirstName(),
                review.getReviewer().getLastName(),
                review.getSeller().getId(),
                review.getProduct() != null ? review.getProduct().getId() : null,
                review.getProduct() != null ? review.getProduct().getName() : null,
                review.getRating(),
                review.getComment(),
                review.getCreatedAt()
            );
        });
    }
    
    /**
     * Calculate and update seller's average rating and total review count.
     */
    private void updateSellerRating(Integer sellerId) {
        List<ReviewEntity> allReviews = reviewRepository.findBySeller_IdOrderByCreatedAtDesc(sellerId);
        
        if (allReviews.isEmpty()) {
            // No reviews, set to 0
            profileRepository.findById(sellerId).ifPresent(profile -> {
                profile.setSellerRating(BigDecimal.ZERO);
                profile.setTotalReviews(0);
                profileRepository.save(profile);
            });
        } else {
            // Calculate average rating
            double average = allReviews.stream()
                .mapToInt(ReviewEntity::getRating)
                .average()
                .orElse(0.0);
            
            BigDecimal avgRating = BigDecimal.valueOf(average)
                .setScale(2, RoundingMode.HALF_UP);
            
            profileRepository.findById(sellerId).ifPresent(profile -> {
                profile.setSellerRating(avgRating);
                profile.setTotalReviews(allReviews.size());
                profileRepository.save(profile);
            });
        }
    }
}

File: frontend/src/services/reviewService.js
import api from './api';

/**
 * Review Service
 * Handles all review/rating-related API calls
 */

/**
 * Create a new review/rating
 * @param {Object} reviewData - Review data
 * @param {number} reviewData.reviewerId - ID of the reviewer (buyer)
 * @param {number} reviewData.sellerId - ID of the seller being reviewed
 * @param {number} reviewData.productId - ID of the product (optional)
 * @param {number} reviewData.orderId - ID of the order
 * @param {number} reviewData.rating - Rating (1-5)
 * @param {string} reviewData.comment - Review text
 * @returns {Promise<Object>} Created review object
 */
export const createReview = async (reviewData) => {
  const response = await api.post('/reviews', {
    reviewer: { id: reviewData.reviewerId },
    seller: { id: reviewData.sellerId },
    product: reviewData.productId ? { id: reviewData.productId } : null,
    order: { id: reviewData.orderId },
    rating: reviewData.rating,
    comment: reviewData.comment
  });
  return response.data;
};

/**
 * Get all reviews for a specific product
 * @param {number} productId - Product ID
 * @returns {Promise<Array>} List of reviews
 */
export const getReviewsByProduct = async (productId) => {
  const response = await api.get(`/reviews/product/${productId}`);
  return response.data;
};

/**
 * Get all reviews received by a seller
 * @param {number} sellerId - Seller profile ID
 * @returns {Promise<Array>} List of reviews
 */
export const getReviewsBySeller = async (sellerId) => {
  const response = await api.get(`/reviews/user/${sellerId}`);
  return response.data;
};

/**
 * Get detailed reviews for a seller with pagination
 * @param {number} sellerId - Seller profile ID
 * @param {number} page - Page number (0-indexed)
 * @param {number} size - Page size
 * @param {string} sortBy - Sort option: 'recent', 'highest', 'lowest'
 * @returns {Promise<Object>} Paginated reviews with content, totalPages, totalElements
 */
export const getDetailedReviewsBySeller = async (sellerId, page = 0, size = 10, sortBy = 'recent') => {
  const response = await api.get(`/reviews/seller/${sellerId}/detailed`, {
    params: { page, size, sortBy }
  });
  return response.data;
};

/**
 * Get all reviews written by a user
 * @param {number} reviewerId - Reviewer profile ID
 * @returns {Promise<Array>} List of reviews
 */
export const getReviewsWritten = async (reviewerId) => {
  const response = await api.get(`/reviews/written/${reviewerId}`);
  return response.data;
};

/**
 * Get a specific review by ID
 * @param {number} reviewId - Review ID
 * @returns {Promise<Object>} Review object
 */
export const getReviewById = async (reviewId) => {
  const response = await api.get(`/reviews/${reviewId}`);
  return response.data;
};

/**
 * Update an existing review
 * @param {number} reviewId - Review ID
 * @param {Object} reviewData - Updated review data (rating, comment)
 * @returns {Promise<Object>} Updated review object
 */
export const updateReview = async (reviewId, reviewData) => {
  const response = await api.put(`/reviews/${reviewId}`, {
    rating: reviewData.rating,
    comment: reviewData.comment
  });
  return response.data;
};

/**
 * Delete a review
 * @param {number} reviewId - Review ID
 * @returns {Promise<void>}
 */
export const deleteReview = async (reviewId) => {
  const response = await api.delete(`/reviews/${reviewId}`);
  return response.data;
};

/**
 * Calculate average rating from reviews array
 * @param {Array} reviews - Array of review objects
 * @returns {number} Average rating (0 if no reviews)
 */
export const calculateAverageRating = (reviews) => {
  if (!reviews || reviews.length === 0) return 0;
  const sum = reviews.reduce((acc, review) => acc + review.rating, 0);
  return (sum / reviews.length).toFixed(1);
};

/**
 * Get rating distribution
 * @param {Array} reviews - Array of review objects
 * @returns {Object} Distribution object {5: count, 4: count, ...}
 */
export const getRatingDistribution = (reviews) => {
  const distribution = { 5: 0, 4: 0, 3: 0, 2: 0, 1: 0 };
  if (!reviews || reviews.length === 0) return distribution;
  
  reviews.forEach(review => {
    if (review.rating >= 1 && review.rating <= 5) {
      distribution[review.rating]++;
    }
  });
  
  return distribution;
};
